dsgrid/config/dimension_associations.py

```diff
@@ -0,0 +1,209 @@
+import itertools
+import logging
+from pathlib import Path
+from typing import Dict
+
+from pyspark.sql.types import StringType
+
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.exceptions import DSGInvalidDimensionAssociation
+from dsgrid.utils.spark import read_dataframe
+
+
+logger = logging.getLogger(__name__)
+
+
+class DimensionAssociations:
+    """Interface to a project's dimension associations"""
+
+    def __init__(self, table):
+        self._table = table
+        if table is None:
+            self._dimensions = set()
+        else:
+            self._dimensions = set((DimensionType.from_column(x) for x in table.columns))
+
+    @classmethod
+    def load(cls, path: Path, association_files):
```

Comment:
@mooneyme I refactored this class to store a single table containing all associations. Please check the joins again. I left your asserts out. I'm uncertain if they are needed. I typically use asserts to guard against coding errors and not bad inputs. If you were guarding against bad input then they need to be added back.

I did not include a join of the project config records because I think it's better to keep the code modular. You can still create that table very easily in `project_registry_manager.py` in the `_check_dataset_base_to_project_base_mappings` method. The main advantage of this approach is that dimension_associations.py can be tested without a project.

You will also see that I left the code that performs project/dataset checks by iterating over pairs of dimensions. As discussed, it might be possible to make this faster by comparing one overall table with the dataset table (`dim_table` in the method above).